data/python.gram

```diff
@@ -171,59 +189,86 @@ class Parser(Parser):
+                    return self.get_invalid_target(target, node.left)
+                return None
+
+            return node
+        elif isinstance(node, (ast.Name, ast.Subscript, ast.Attribute)):
+            return None
+        else:
+            return node
+
     def set_expr_context(self, node, context):
         """Set the context (Load, Store, Del) of an ast node."""
         node.ctx = context
         return node
 
-    def ensure_real(self, number_str: str):
-        number = ast.literal_eval(number_str)
-        if type(number) is complex:
-            self.raise_syntax_error("real number required in complex literal")
-        return number
+    def ensure_real(self, number: ast.Constant):
+        value = ast.literal_eval(number.string)
+        if type(value) is complex:
+            self.raise_syntax_error_known_location("real number required in complex literal", number)
+        return value
 
-    def ensure_imaginary(self, number_str: str):
-        number = ast.literal_eval(number_str)
-        if type(number) is not complex:
-            self.raise_syntax_error("imaginary number required in complex literal")
-        return number
+    def ensure_imaginary(self, number:  ast.Constant):
+        value = ast.literal_eval(number.string)
+        if type(value) is not complex:
+            self.raise_syntax_error_known_location("imaginary number required in complex literal", number)
+        return value
 
     def generate_ast_for_string(self, tokens):
         """Generate AST nodes for strings."""
-        err_msg = ''
-        line = 1
+        err = None
+        line_offset = tokens[0].start[0]
+        line = line_offset
         col_offset = 0
-        source = ''
+        source = "(\\n"
         for t in tokens:
             n_line = t.start[0] - line
             if n_line:
                 col_offset = 0
-            source += """\n""" * n_line + ' ' * (t.start[1] - col_offset) + t.string
+            source += """\\n""" * n_line + ' ' * (t.start[1] - col_offset) + t.string
             line, col_offset = t.end
-        if source[0] == ' ':
-            source = '(' + source[1:]
-        else:
-            source = '(' + source
-        source += ')'
+        source += "\\n)"
         try:
             m = ast.parse(source)
         except SyntaxError as e:
-            err_msg = e.args[0]
-            # Identify the line at which the error occurred to get a more
-            # accurate line number
-            for t in tokens:
-                try:
-                    m = ast.parse(t.string)
-                except SyntaxError:
-                    break
+            err = e
 
         # Avoid getting a triple nesting in the error report that does not
         # bring anything relevant to the traceback.
-        if err_msg:
-            self.raise_syntax_error_known_location(err_msg, t)
-            raise self._exception
-
-        return m.body[0].value
+        if err is not None:
```

Comment:
This keeps the exception alive more than it should, and this will delay the destruction of the frame due to it forming a cycle with the frame. Let's make sure we delete de reference explicitly when we don't need it anymore by adding `del err` or by changing this code to capture only the message (as it was previously).

Reply:
This would mean losing the exact matching to CPython (which I worked hard to get). I would prefer adding an explicit delete even though I do not see much difference since we can only delete it right before raising.

Reply:
> delete it right before raising.

That makes all the difference because it avoids the cycle that will be propagated as the frame in the exception bubbles up. This is anyway a small detail, so don't worry too much about it if you want to leave as is :)